Tl_dev

theo-long opened the pull request:
Get the MNIST LMC basic example notebook working, and some edits to the models to remove softmax layers

char-tan reviewed (approved):
Nice work - looks good to me! I'll hold off on merging to let @sarmxzh and @LRudL take a look.

My one observation (I don't think this needs changing for this PR - as they only need MLP working atm)

I think CosineAnnealingLR needs to have T_max = iterations_per_epoch * num_epochs
It's currently set to T_max = num_epochs

Then we pass optimizer.step() into train() and call after every batch has been processed (currently doing it once per epoch). I think the way it's setup now will work - but just mean the LR steps are really big at the end of each epoch.

Lastly, to use e.g OneCycleLR (to have warmup) we'll need a way of passing kwargs into the class initialisation - or alternatively just have an option 'onecycle' that we can pass in

Once this is merged I'd be happy to do these scheduler bits.

char-tan commented:
Actually - I've just run the notebook on colab - ran fine up until the plotting but getting some errors here. The first I understand, the second less so

![Screenshot from 2022-12-31 13-22-31](https://user-images.githubusercontent.com/76977995/210138270-5ed0e372-2bbc-4441-8898-52a0c00eb9fa.png)

![Screenshot from 2022-12-31 13-23-13](https://user-images.githubusercontent.com/76977995/210138274-97affc25-82ae-4e76-a046-70630f2b29f7.png)

char-tan commented:
(sorry for images) - is there a good way to paste error messages?

theo-long commented:
@char-tan For the CosineAnnealingLR I agree - schedule.step() call needs to be change. If you're happy to fix that bit that would be v. helpful.

For the two errors - one was because I hardcoded the "mps" device since I'm running it locally. I've changed it to now use get_device so that error should be gone. For the other error - did you change n_points to 3 when you ran it by any chance? The number of lambdas needs to be the same as the n_points - I refactored it so there's a variable n_points that is used for both the interpolation and the plotting.

char-tan reviewed (approved):
Looks good!
